### ISSUE TITLE
moveTo property needs a dropdown menu showing list of Views

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#1121]
@jhpedemonte, @childsb

The moveto property should have a dropdown that shows a list of IDs for View, ScrollableView and SwapView widgets, versus having users have to remember the IDs they used.

**Comment by JonFerraiolo** [comment]
See also #247, #1043, #1044, #249

**Comment by JonFerraiolo** [comment]
Changed to Priority-Low like other issues that request drop-down menus (e.g., id and class properties)

**Comment by JonFerraiolo** [comment]
Pushing to M5

**Comment by jhpedemonte** [comment]
I'm thinking we can make this extensible via a bit of metadata, such that other props can make use of it.

Something where we can set filters in the metadata value, in order to narrow down what is shown. We can maybe use CSS selectors -- for example:

```
[data-dojo-type="dojox.mobile.View"], [data-dojo-type="dojox.mobile.ScrollableView"], [data-dojo-type="dojox.mobile.SwapView"]
```

And we also need a way to specify what to list.  In this case, it would be the IDs of matching widgets.

**Comment by jhpedemonte** [comment]
-> M6/Med

**Comment by doronrosenberg** [comment]
Since moveTo is used in several places, is there a way to define the needed metadata only once (the view/scrollableview/etc part) and reference it in the oam.json files?

**Comment by JonFerraiolo** [comment]
Short answer: sorry, no, have to repeat things, no way to reference a shared definition.

Longer answer: During the OpenAjax metadata standards discussion, we talked about an "include" feature  and had actually spec'd the feature and mentioned it in the Futures appendix (http://www.openajax.org/member/wiki/OpenAjax_Metadata_Specification_Futures#.3Cinclude.3E), but didn't include it in the 1.0 (and only) version of the spec because: (1) maybe XInclude or other standard should be used instead, (2) no one had implemented it yet. For us to enhance the OAM format we would need a hugely compelling reason. So far we have managed to use it without extensions, except for inventing a JSON encoding (also mentioned in Futures appendix: http://www.openajax.org/member/wiki/OpenAjax_Metadata_Specification_Futures#Standard_JSON_representation_of_OpenAjax_Metadata).

**Comment by doronrosenberg** [comment]
question - do we want the user to be able to enter manual ids as well as dropdown values or just choose from a dropdown?

**Comment by JonFerraiolo** [comment]
user should be able to enter manual ids, too (i.e., combobox instead of just a select)

**Comment by doronrosenberg** [comment]
So right now I am calling the property dropdownQueryValues:

``` javascript
        "moveTo": {
            "datatype": "string",
            "dropdownQueryValues": ["[data-dojo-type='dojox.mobile.View']", "[data-dojo-type='dojox.mobile.SwapView']", "[data-dojo-type='dojox.mobile.ScrollableView']"]
        },
```

Looking at OpenAjax I don't see an appropriate name, does the spec allow "foreign" properties?

**Comment by JonFerraiolo** [comment]
Sorry, but I expect this new feature is going to take more time to talk about than implement.

Over multiple months early in the project, Javier and I went over the policy question about whether Maqetta should extend the OAM format, and our conclusion was "no" and instead if OAM doesn't provide what we need, then we should use a different approach to solving the problem at hand. Typically, the "different approach" consists of:
- Adding addition Maqetta-specific metadata into the packages.json or widgets.json file for the particular widget
- Doing something with helpers

As a background, where we draw the line for what sort of info belongs in the OAM file versus the packages.json/widgets.json file is this rule: generic widget metadata that would make as much sense to other authoring tools (e.g., Dreamweaver, which also supports OAM), and can be expressed in OAM without extension, belongs in the OAM file. Anything else (i.e., Maqetta-specific metadata or extensions to OAM) should go into packages.json/widgets.json.

IMO, this particular problem isn't clear-cut about what's the best approach to use. I would strongly vote against extending OAM and look at one of the two bullets above. In general, I'll defer to @jhpedemonte about whether to go for one of the two bullets above or something else, and figuring out the details, but here are some thoughts after thinking about this for a bit:
- In the OAM file, just leave moveTo as a single string
- Invent one or more new properties for packages.json/widgets.json that express what is needed, but attempt to make this metadata generic and reusable. 
- To achieve genericism, while still addressing this particular problem, we probably need to also provide a new helper function.

What I'm thinking is that there is a new construct in packages.json/widgets.json that provides generic "OAM property overrides", and one of the override features that is available is something to say that a particular property (e.g., 'moveTo') should present a particular UI (e.g., comboBox), and then the core code would pull the list of choices for the comboBox via a well-defined helper function (helper.getPropertyValueOptions).

**Comment by doronrosenberg** [comment]
I think the best choice would be via a helper, a preProcessProperty perhaps that would be called for each property and in this case replace the type with comboEdit and add the values from the DOM.

**Comment by JonFerraiolo** [comment]
Sounds like a reasonable direction. Why not make it into a generic `metadataPostProcess` helper where you pass a pointer to the entire metadata JavaScript object (which holds the JavaScript objects corresponding to the OAM JSON data), and then the helper can add/remove/modify values on the metadata object as is needed, including Maqetta-specific property metadata such as what is needed here.

**Comment by JonFerraiolo** [comment]
Wait, declarative is better than procedural. Maybe we could provide a metadata mixin mechanism within packages.json/widgets.json, which under the hood would end up invoking dojo.mixin. This would allow adding Maqetta-specific things to the OAM data.

**Comment by doronrosenberg** [comment]
The benefit of doing it in JS is that it can do more, but right now there isn't a need for more than doing a mixin.

A more complex example could be metadata processing based on which mobile device is set.  Perhaps different properties depending if a tablet is chosen vs a mobile phone.  I believe there is a bug to disable width/height on mobile Views if we are in flow layout which would be another example.

**Comment by JonFerraiolo** [comment]
Yes, whatever we do, we should architect things so that there is a mechanism to invoke a JavaScript helper for the reasons you say above.

**Comment by doronrosenberg** [comment]
Proposal:

`metadataPostProcess` - helper method that gets called **once** when the metadata is first loaded from the OAM file.

`propertyPostProcess` - helper method that gets called **every time** a property is displayed in the properties palette.  Allows for dynamic changes before the property data is processed.

Entry in packages/widgets.json called oamOverride that gets mixed in into the OAM metadata **once** when it is first loaded.  Happens probably before metadataPostProcess gets called.

**Comment by jhpedemonte** [comment]
I'd prefer to do this in a **declarative** fashion.  Something like this, if generic enough, could also be used elsewhere -- don't we have another area where we want to display a drop down (instead of a text area) based on the widgets on the page?

Also, a declarative approach makes it easier for other developers to use.

**Comment by doronrosenberg** [comment]
That would be the .json change for packages/widgets.json, which would mixin with the OAM json.  We could easily push the JS stuff into another issue in case we ever need more powerful functionality and keep this issue to using declarative changes.

**Comment by doronrosenberg** [comment]
Turns out packages/widgets.json already supports mixing in metadata into the OAM data using the "metadata" property.

**Comment by JonFerraiolo** [comment]
Good catch! Was added by @jhpedemonte in this commit: https://github.com/maqetta/maqetta/commit/256aaee168d757874cb20798e125c96bebb47556#davinci.dojo_1_7/WebContent/metadata/widgets.json

**Comment by jhpedemonte** [comment]
Hey-oh!  I knew I was good for something!
